File: algorithms/divide-and-conquer/strassen-matrix-multiplication/StrassenMatrixMultiplication.py
class MatrixMultiplication:
    # given 2 matrices A and B: return A + B
    @staticmethod
    def add(A, B):
        return [
            [A[i][j] + B[i][j] for j in range(len(A[0]))] for i in range(len(A))
        ]

    # given 2 matrices A and B: return A - B
    @staticmethod
    def subtract(A, B):
        return [
            [A[i][j] - B[i][j] for j in range(len(A[0]))] for i in range(len(A))
        ]

    # given 2 matrices A and B: return A X B (Strassen)
    @staticmethod
    def strassen(A, B):
        n = len(A)

        # base case: single element multiplication
        if n == 1:
            return [[A[0][0] * B[0][0]]]

        k = n // 2

        # divide matrices into 4 parts each
        A11 = [row[:k] for row in A[:k]]
        A12 = [row[k:] for row in A[:k]]
        A21 = [row[:k] for row in A[k:]]
        A22 = [row[k:] for row in A[k:]]

        B11 = [row[:k] for row in B[:k]]
        B12 = [row[k:] for row in B[k:]]
        B21 = [row[:k] for row in B[k:]]
        B22 = [row[k:] for row in B[k:]]

        # compute the 7 products (core idea of Strassen)
        P1 = MatrixMultiplication.strassen(
            A11, MatrixMultiplication.subtract(B12, B22)
        )
        P2 = MatrixMultiplication.strassen(
            MatrixMultiplication.add(A11, A12), B22
        )
        P3 = MatrixMultiplication.strassen(
            MatrixMultiplication.add(A21, A22), B11
        )
        P4 = MatrixMultiplication.strassen(
            A22, MatrixMultiplication.subtract(B21, B11)
        )
        P5 = MatrixMultiplication.strassen(
            MatrixMultiplication.add(A11, A22),
            MatrixMultiplication.add(B11, B22),
        )
        P6 = MatrixMultiplication.strassen(
            MatrixMultiplication.subtract(A12, A22),
            MatrixMultiplication.add(B21, B22),
        )
        P7 = MatrixMultiplication.strassen(
            MatrixMultiplication.subtract(A11, A21),
            MatrixMultiplication.add(B11, B12),
        )

        # merge partial results into final quadrants
        C11 = MatrixMultiplication.add(
            MatrixMultiplication.subtract(
                MatrixMultiplication.add(P5, P4), P2
            ),
            P6,
        )
        C12 = MatrixMultiplication.add(P1, P2)
        C21 = MatrixMultiplication.add(P3, P4)
        C22 = MatrixMultiplication.subtract(
            MatrixMultiplication.subtract(
                MatrixMultiplication.add(P1, P5), P3
            ),
            P7,
        )

        C = []

        # place each quadrant into result matrix
        for i in range(k):
            C.append(C11[i] + C12[i])
        for i in range(k):
            C.append(C21[i] + C22[i])

        return C

File: algorithms/divide-and-conquer/strassen-matrix-multiplication/StrassenMatrixMultiplication.rs
fn add(a: &Vec<Vec<i32>>, b: &Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let n = a.len();
    let mut result = vec![vec![0; n]; n];

    for i in 0..n {
        for j in 0..n {
            result[i][j] = a[i][j] + b[i][j];
        }
    }

    result
}

fn subtract(a: &Vec<Vec<i32>>, b: &Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let n = a.len();
    let mut result = vec![vec![0; n]; n];

    for i in 0..n {
        for j in 0..n {
            result[i][j] = a[i][j] - b[i][j];
        }
    }

    result
}

fn strassen(a: &Vec<Vec<i32>>, b: &Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let n = a.len();

    if n == 1 {
        return vec![vec![a[0][0] * b[0][0]]];
    }

    let k = n / 2;

    let mut a11 = vec![vec![0; k]; k];
    let mut a12 = vec![vec![0; k]; k];
    let mut a21 = vec![vec![0; k]; k];
    let mut a22 = vec![vec![0; k]; k];

    let mut b11 = vec![vec![0; k]; k];
    let mut b12 = vec![vec![0; k]; k];
    let mut b21 = vec![vec![0; k]; k];
    let mut b22 = vec![vec![0; k]; k];

    for i in 0..k {
        for j in 0..k {
            a11[i][j] = a[i][j];
            a12[i][j] = a[i][j + k];
            a21[i][j] = a[i + k][j];
            a22[i][j] = a[i + k][j + k];

            b11[i][j] = b[i][j];
            b12[i][j] = b[i][j + k];
            b21[i][j] = b[i + k][j];
            b22[i][j] = b[i + k][j + k];
        }
    }

    let p1 = strassen(&a11, &subtract(&b12, &b22));
    let p2 = strassen(&add(&a11, &a12), &b22);
    let p3 = strassen(&add(&a21, &a22), &b11);
    let p4 = strassen(&a22, &subtract(&b21, &b11));
    let p5 = strassen(&add(&a11, &a22), &add(&b11, &b22));
    let p6 = strassen(&subtract(&a12, &a22), &add(&b21, &b22));
    let p7 = strassen(&subtract(&a11, &a21), &add(&b11, &b12));

    let c11 = add(&subtract(&add(&p5, &p4), &p2), &p6);
    let c12 = add(&p1, &p2);
    let c21 = add(&p3, &p4);
    let c22 = subtract(&subtract(&add(&p1, &p5), &p3), &p7);

    let mut c = vec![vec![0; n]; n];

    for i in 0..k {
        for j in 0..k {
            c[i][j] = c11[i][j];
            c[i][j + k] = c12[i][j];
            c[i + k][j] = c21[i][j];
            c[i + k][j + k] = c22[i][j];
        }
    }

    c
}
